### PR TITLE
Enh #7488: Added `StringHelper::explode` to perform explode with trimmin...

### PR DESCRIPTION
**Pull request by SilverFire**
...g and skipping of empty elements

### CODE REVIEW

**Comment by creocoder** [comment]
1) Code style is wrong
2) You do not need `array_push` for simple `$array[] = ...;`

**Comment by SilverFire** [comment]
Sorry for my bad, it's my first step to contribute Yii2, so hope on your patience

**Comment by creocoder** [comment]
> Sorry for my bad, it's my first step to contribute Yii2, so hope on your patience.

Its fine, you are always welcome :)

> Now it's better?

Now its amazing! Just add your nickname right after change line inside `changelog.md`.

**Comment by creocoder** [comment]
@SilverFire There is one more thing about algorithm

```php
if ($skipEmpty) {
    if (static::byteLength($value)) {
        $result[] = $value;
    }
} else {
    $result[] = $value;
}
```

You can refactor it like:

```php
if ($skipEmpty && !static::byteLength($value)) {
    continue;
}
$result[] = $value;
```

Its shorter and more clean.

**Comment by SilverFire** [comment]
@creocoder, you're absolutely right. Here is a new push.
What do you think about this small enhancement? Can it be useful?

**Comment by creocoder** [comment]
@SilverFire Yes, its useful. For example same task i was solved inside mine https://github.com/creocoder/yii2-taggable/blob/master/src/TaggableBehavior.php#L216 . With that method it would be more clean.

File: framework/helpers/BaseStringHelper.php [hunk 1]
```diff
@@ -234,4 +234,26 @@ public static function endsWith($string, $with, $caseSensitive = true)
...
+     * @param boolean $trim Whether to trim each element
+     * @param boolean $skipEmpty Whether to skip empty strings between delimiters
+     * @return array
+     */
+    public static function explode($string, $delimiter = ',', $trim = true, $skipEmpty = false)
+    {
+        $result = [];
+        foreach (explode($delimiter, $string) as $value) {
+            $value = $trim ? trim($value) : $value;
+            if ($skipEmpty && !static::byteLength($value)) {
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php]: `!static::byteLength($value)` could be replaced with `$value === ''`.

Comment (creocoder) [inline framework/helpers/BaseStringHelper.php]: @samdark Should be replaced in other `StringHelper` methods than.

Comment (creocoder) [inline framework/helpers/BaseStringHelper.php]: Also probably loop can be replaced with `array_*` methods.

Comment (SilverFire) [inline framework/helpers/BaseStringHelper.php]: @samdark - yes, you're right. Replaced.
@creocoder - I thought about it yesterday, but didn't found the way to implement with `array_*` methods.

`array_map`, `array_walk` allows to change values (useful if we need to trim only), but does not allow to remove element from an array
`array_filter` allows to remove element from an array, but does not allow to change element's value

I doubt that using both of them will be faster than `foreach`. Am I missing something?

File: framework/helpers/BaseStringHelper.php [hunk 2]
```diff
@@ -234,4 +234,26 @@ public static function endsWith($string, $with, $caseSensitive = true)
...
+     * @param string $delimiter
+     * @param boolean $trim Whether to trim each element
+     * @param boolean $skipEmpty Whether to skip empty strings between delimiters
+     * @return array
+     */
+    public static function explode($string, $delimiter = ',', $trim = true, $skipEmpty = false)
+    {
+        $result = [];
+        foreach (explode($delimiter, $string) as $value) {
+            $value = $trim ? trim($value) : $value;
```
Comment (nineinchnick) [inline framework/helpers/BaseStringHelper.php]: Can `$trim` be a mixed value? False for no trim, true for normal trim and a string passed as second argument:
~~~php
if ($trim) {
    $value = $trim === true ? trim($value) : trim($value, $trim);
}
~~~

**Comment by nineinchnick** [comment]
I usually combine `array_map` with `array_filter` like so:

~~~php
$value = array_filter(array_map('trim', explode(',', $value)));
~~~

Sometimes I add another `array_map` after `array_filter` to validate numerics:

~~~php
$value = array_map('intval', array_filter(array_map('trim', explode(',', $value))));
~~~

But it's not really more readable than a simple `foreach` and I didn't do any benchmarks, so I don't know if it's faster.

**Comment by nineinchnick** [comment]
I've done [a simple benchmark](https://gist.github.com/nineinchnick/ab6a3ccc938eeec238f5):

                original * 100000: 1.6924879550934
                 array_* * 100000: 1.4656081199646
    array_* without vars * 100000: 1.4615139961243

**Comment by SilverFire** [comment]
@nineinchnick the second looks good, but my try of your benchmark shows another:

PHP 5.5.22

```
                original * 100000: 0.55539298057556
                 array_* * 100000: 0.69772720336914
    array_* without vars * 100000: 0.63055896759033
```


PHP 5.6.6

```
                original * 100000: 0.64687585830688
                 array_* * 100000: 0.73420190811157
    array_* without vars * 100000: 0.65790700912476
```


http://3v4l.org/cXsaN

**Comment by nineinchnick** [comment]
I've updated the benchmark a bit and [run it myself](http://3v4l.org/ZLh9h).

On my computer (i5, PHP 5.6.6) `array_*` versions are faster but most other machines it's the opposite. In conculsion I'd use the `foreach` version as it is more readable.

File: framework/helpers/BaseStringHelper.php [hunk 3]
```diff
@@ -234,4 +234,26 @@ public static function endsWith($string, $with, $caseSensitive = true)
...
+     * @param string $string
+     * @param string $delimiter
+     * @param boolean $trim Whether to trim each element
+     * @param boolean $skipEmpty Whether to skip empty strings between delimiters
+     * @return array
+     */
+    public static function explode($string, $delimiter = ',', $trim = true, $skipEmpty = false)
+    {
+        $result = [];
+        foreach (explode($delimiter, $string) as $value) {
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php]: Move explode out of `foreach`. http://3v4l.org/SV9mJ

**Comment by nineinchnick** [comment]
I've updated [my benchmark](https://gist.github.com/nineinchnick/ab6a3ccc938eeec238f5) with a version that allows to pass a callable or a string as `$trim`. This allows more flexibility in filtering. Also, when trimming is disabled, the `array_*` version are way (2x) faster.

**Comment by nineinchnick** [comment]
[Another update](https://gist.github.com/nineinchnick/ab6a3ccc938eeec238f5) with more argument variants (trim/notrim and empty/noempty) tested at the same time.

**Comment by creocoder** [comment]
I think we need go with `array_*` methods, its little faster on some test, its little slower on one tests.

> On my computer (i5, PHP 5.6.6) array_* versions are faster but most other machines it's the opposite. In conculsion I'd use the foreach version as it is more readable.

Its very controversial conclusion. Its shorter and faster in most tests. I do not think anybody will open `StringHelper`, move to `explode` and will spens hours on meditating on its implementation. Readability sometimes make sence. But in general readability does not mean simple for php noobs, in second readability should never affect perfomance especially inside frameworks.

**Comment by nineinchnick** [comment]
@creocoder I agree after studying the benchark results for a while.

**Comment by SilverFire** [comment]
Also agree, that `array_*` functions look good. 

After applying `array_filter` the array's keys may be nonsequential like this:

```

1) yiiunit\framework\helpers\StringHelperTest::testExplode
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => 'Disable'
     1 => '  trim  '
-    2 => 'here but ignore empty'
+    4 => 'here but ignore empty'
 )
```

So need to wrap a result of `array_filter` with `array_values` to have sequential keys of an array.
This modification does not decrease performance of function.

**Comment by samdark** [comment]
#7488

File: framework/helpers/BaseStringHelper.php [hunk 4]
```diff
@@ -234,4 +234,28 @@ public static function endsWith($string, $with, $caseSensitive = true)
...
         }
     }
+
+    /**
+     * Explodes string to array, optionally trims values, skip empty
+     *
+     * @param string $string
+     * @param string $delimiter Default is ','
+     * @param mixed $trim Whether to trim each element. Can be:
+     *                    true - to normal trim
```
Comment (nineinchnick) [inline framework/helpers/BaseStringHelper.php]: Should be: `boolean - to normal trim`

**Comment by samdark** [comment]
Merged with some adjustments. Thanks!

**Pull request closed without merging**